packages/config-loader/src/lib/schema/compile.ts

```diff
@@ -35,6 +35,48 @@ import {
 export function compileConfigSchemas(
   schemas: ConfigSchemaPackageEntry[],
 ): ValidationFunc {
+  const { schema: merged, parser: ajv, visibilityByPath } = mergeConfigSchemas(
+    schemas,
+  );
+
+  const validate = ajv.compile(merged);
+
+  return configs => {
+    const config = ConfigReader.fromConfigs(configs).get();
+
+    visibilityByPath.clear();
+
+    const valid = validate(config);
+    if (!valid) {
+      const errors = validate.errors ?? [];
+      return {
+        errors: errors.map(({ dataPath, message, params }) => {
+          const paramStr = Object.entries(params)
+            .map(([name, value]) => `${name}=${value}`)
+            .join(' ');
+          return `Config ${message || ''} { ${paramStr} } at ${dataPath}`;
+        }),
+        visibilityByPath: new Map(),
+      };
+    }
+
+    return {
+      visibilityByPath: new Map(visibilityByPath),
+    };
+  };
+}
+
+/**
+ * Given a list of configuration schemas from packages, merge them
+ * into a single json schema.
+ */
+export function mergeConfigSchemas(
```

Comment:
I think this contains quite a lot more than it should, it should really just be the `mergeAllOf` bit. We could also make it take an array of schema so we don't need to export the `ConfigSchemaPackageEntry` type, as that one's pretty internal. Basically ending up with `(schemas: JSONSchema[]) => JSONSchema`. In the CLI already verify the schema when loading it, so no need to make sure that's logic as part of merging the schema back together again in the CLI.

Reply:
Hah, indeed. I somehow gathered that the Ajv compiler was used in the merge operation. Don't know how I gathered that, now when I look at it again :p 
Refactored and cleaned up.